pkg/controllers/nodeoverlay/store.go

```diff
@@ -0,0 +1,287 @@
+	lowestWeight            *int32
+}
+
+type InstanceTypeUpdate struct {
+	Price    map[string]*PriceUpdate
+	Capacity *CapacityUpdate
+}
+
+type InstanceTypeStore struct {
+	updates map[string]map[string]*InstanceTypeUpdate
+	mu      sync.RWMutex
+}
+
+func NewInstanceTypeStore() *InstanceTypeStore {
+	return &InstanceTypeStore{
+		updates: map[string]map[string]*InstanceTypeUpdate{},
+	}
+}
+
+func (s *InstanceTypeStore) UpdateStore(updatedStore map[string]map[string]*InstanceTypeUpdate) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.updates = map[string]map[string]*InstanceTypeUpdate{}
+	for nodePoolName, v := range updatedStore {
+		s.updates[nodePoolName] = map[string]*InstanceTypeUpdate{}
+		for overlayName, itUpdate := range v {
+			s.updates[nodePoolName][overlayName] = &InstanceTypeUpdate{
+				Price:    itUpdate.Price,
+				Capacity: itUpdate.Capacity,
+			}
+		}
+	}
+}
+
+func (s *InstanceTypeStore) FindInstanceTypesUpdate(nodePoolName string, it *cloudprovider.InstanceType) (*InstanceTypeUpdate, bool) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	instanceTypeList, foundNodePool := s.updates[nodePoolName]
+	if !foundNodePool {
+		return nil, false
+	}
+	instanceTypeUpdate, foundInstanceTypeUpdate := instanceTypeList[it.Name]
+	return instanceTypeUpdate, foundInstanceTypeUpdate
+}
+
+// updateInstanceTypeCapacity add a new Capacity overlay update to the associated instance type.
+// This add an Capacity update to the store once after it has been validated.
+func (s *InstanceTypeStore) updateInstanceTypeCapacity(nodePoolName string, instanceTypeName string, nodeOverlay v1alpha1.NodeOverlay) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if nodeOverlay.Spec.Capacity == nil {
+		return
+	}
+
+	_, foundNodePool := s.updates[nodePoolName]
+	if !foundNodePool {
+		s.updates[nodePoolName] = map[string]*InstanceTypeUpdate{}
+	}
+	_, foundInstanceTypeUpdate := s.updates[nodePoolName][instanceTypeName]
+	if !foundInstanceTypeUpdate {
+		s.updates[nodePoolName][instanceTypeName] = &InstanceTypeUpdate{Price: map[string]*PriceUpdate{}, Capacity: &CapacityUpdate{OverlayUpdate: corev1.ResourceList{}}}
+	}
+
+	if s.updates[nodePoolName][instanceTypeName].Capacity == nil {
+		s.updates[nodePoolName][instanceTypeName].Capacity = &CapacityUpdate{
+			OverlayUpdate:           nodeOverlay.Spec.Capacity,
+			lowestWeightOverlayName: nodeOverlay.Name,
+			lowestWeight:            nodeOverlay.Spec.Weight,
+		}
+	} else {
+		for resource, qun := range nodeOverlay.Spec.Capacity {
+			s.updates[nodePoolName][instanceTypeName].Capacity.OverlayUpdate[resource] = qun
+		}
+		s.updates[nodePoolName][instanceTypeName].Capacity.lowestWeight = nodeOverlay.Spec.Weight
+		s.updates[nodePoolName][instanceTypeName].Capacity.lowestWeightOverlayName = nodeOverlay.Name
+	}
+
+}
+
+func (s *InstanceTypeStore) isCapacityUpdateConflicting(nodePoolName string, instanceTypeName string, nodeOverlay v1alpha1.NodeOverlay) (string, bool) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
```

Comment:
It shouldn't matter too much in practice, but we never need a mutex here since the store is guaranteed to not be accessed concurrently while being generated. This makes me wonder if there's a bit of an abstraction issue (e.g. maybe we shouldn't be passing around the store directly, but make the controller implement an interface like this and have a mutex to lock the store for these functions):

```go
type InstanceOverlayProvider interface {
  func ApplyOverlay(nodePool, instanceType) instanceType
}
```

Note that the interface is just for demonstrative purposes - we don't literally need to make this type here, if we wanted to use an interface it should be created by the callee anyway.

Reply:
> It shouldn't matter too much in practice, but we never need a mutex here since the store is guaranteed to not be accessed concurrently while being generated. 

Can you elaborate more? It's still possible to access the store while it's still in the process of being updated. There will not be anything blocking it from being updated. I have made several change, I'm not sure this comment still applies very much